fsw/cfe-core/src/sb/cfe_sb_api.c

```diff
@@ -1651,12 +1651,23 @@ uint32  CFE_SB_GetLastSenderId(CFE_SB_SenderId_t **Ptr,CFE_SB_PipeId_t  PipeId)
     /* Get ptr to buffer descriptor for the last msg received on the given pipe */
     Ptr2BufDescriptor = CFE_SB.PipeTbl[PipeId].CurrentBuff;
 
-    /* Set the receivers pointer to the adr of 'Sender' struct in buf descriptor */
-    *Ptr = (CFE_SB_SenderId_t *) &Ptr2BufDescriptor -> Sender;
-
-    CFE_SB_UnlockSharedData(__func__,__LINE__);
-
-    return CFE_SUCCESS;
+    if ( Ptr2BufDescriptor == NULL  )
+    {
+        *Ptr = NULL;
+        CFE_SB.PipeTbl[PipeId].LastSender = CFE_SB_INVALID_MSG_ID;
+        CFE_EVS_SendEventWithAppID(CFE_SB_NO_LAST_MESG_RECV_EID,CFE_EVS_EventType_ERROR,CFE_SB.AppId,
+              "SB GetLastSender Err:CurrentBuff not initialized - No last sender, Pipe=%d,App=%s",
+              (int)PipeId,CFE_SB_GetAppTskName(TskId,FullName));
+        CFE_SB_UnlockSharedData(__func__,__LINE__);
+        return CFE_SB_NO_MSG_RECV; 
+    }
+    else
+    {
+        /* Set the receivers pointer to the adr of 'Sender' struct in buf descriptor */
+        *Ptr = (CFE_SB_SenderId_t *) &Ptr2BufDescriptor -> Sender;
+        CFE_SB_UnlockSharedData(__func__,__LINE__);
+        return CFE_SUCCESS;
+    }
```

Comment:
Actually, this API is fundamentally broken.  I had never really looked at it before.

The content of `CFE_SB.PipeTbl[PipeId].CurrentBuff` is shared, and needs to be locked while accessing it.  (which it is, within this function).  But then it is gettng a direct pointer to the data within the descriptor, unlocking, and returning that pointer for the application to access.  This is unsafe - that data can be overwritten at any time after the unlock.

Rather, this should take a pointer to the `CFE_SB_SenderId_t` buffer (not a pointer-to-pointer) and copy the data, then unlock.

Of course, this changes the API, and makes it incompatible.

One possible mitigation to make it backward compatible is to have a separate buffer in the PipeTbl struct holding another copy, then copy the data into that and return a pointer to it.  That way at least the returned pointer would remain valid until the next call to the same function on the same pipe (sort of like the way the posix `readdir` issue was mitigated).

Reply:
See #608

Reply:
Agree with the general philosophy.  I just wanted to point out that the documentation does state that *Ptr is only valid until the next call to CFE_SB_RcvMsg for the same pipe.